### ISSUE TITLE
`WP_CLI::add_command` doesn't work

### ISSUE THREAD

**Issue by miya0001** [wp-cli/wp-cli#4084]
I installed a following package on newest nightly version of the WP-CLI.
https://github.com/vccw-team/wp-cli-scaffold-movefile

```
$ wp package install vccw/wp-cli-scaffold-movefile:@stable
```

My package was installed as expected, but it wasn't registered.

```
$ wp help scaffold movefile
Error: 'scaffold movefile' is not a registered wp command.
```

I tried it with newest nightly build.

```
$ wp --info
PHP binary:	/usr/bin/php7.0
PHP version:	7.0.8-0ubuntu0.16.04.3
php.ini used:	/etc/php/7.0/cli/php.ini
WP-CLI root dir:	phar://wp-cli.phar
WP-CLI packages dir:	/home/vagrant/.wp-cli/packages/
WP-CLI global config:	/home/vagrant/.wp-cli/config.yml
WP-CLI project config:	
WP-CLI version:	1.2.0-alpha-a0d0f94
```

But it is working on `1.2.0-alpha-3182ac4`

**Comment by danielbachhuber** [comment]
> My package was installed as expected, but it wasn't registered.

Just to clarify, does the package work when you try `wp scaffold movefile` ?

**Comment by miya0001** [comment]
It isn't working.

```
$ wp scaffold movefile
Error: 'movefile' is not a registered subcommand of 'scaffold'. See 'wp help scaffold'.
```

**Comment by danielbachhuber** [comment]
Ah.

The immediate problem is that you need to replicate https://github.com/wp-cli/scaffold-package-command/pull/104 and https://github.com/wp-cli/scaffold-package-command/pull/105

However, I'd like to keep this issue open because I want @schlessera's input on how we should handle this problem generally.

It seems possible other commands registered to existing command namespaces will experience this problem of unexpectedly no longer working.

**Comment by schlessera** [comment]
The obvious solution would be to throw an error when you try to add a sub-command for a parent that is not known yet.

However, I wonder if a smarter approach is not preferable since we now have the additional hooks. In such a case, we could run a new `add_command()` call that gets hooked to the `after_add_command_<command>` for the parent command. If the parent gets added later, it will trigger adding of the child automatically. Otherwise, nothing will happen, as the action is never triggered.

**Comment by danielbachhuber** [comment]
> The obvious solution would be to throw an error when you try to add a sub-command for a parent that is not known yet.

I'm more keen to find a solution that doesn't break backwards compatibility. Ideas for this?

**Comment by schlessera** [comment]
I'll add an automated dependency resolution mechanism to get around this, as discussed in Slack:

```
schlessera [6:35 PM] 
https://github.com/wp-cli/wp-cli/issues/4084 (edited)

[6:36] 
I had proposed an approach that might automatically fix dependency resolution issues with commands (the 2nd option in my comment).

[6:36] 
It sounds a bit too easy in my mind, so I wanted to go through this to make sure I'm not missing something obvious.

danielbachhuber [6:37 PM] 
oh, now I understand your comment

schlessera [6:37 PM] 
The way I think we can solve this is to make use of the new hook we added from within the `add_command()`. The basic approach would be that the `add_command()` is deferred by the fact that it hooks itself up to the parent command.

danielbachhuber [6:38 PM] 
deferring command registration could work, but we'd still run into race conditions potentially

[6:39] 
if `WP_CLI::add_command( 'scaffold package' )` is still hooked before `WP_CLI::add_command( 'scaffold' )`, then the latter would be registered over the former

[6:39] 
one approach would be to special-case bundled commands to always register before package commands

[6:39] 
er

schlessera [6:39 PM] 
No, if `WP_CLI::add_command( 'scaffold package' )` runs too early, it would hook itself up to `after_add_command:scaffold`.

danielbachhuber [6:40 PM] 
what if `after_add_command:scaffold`  never fires?

schlessera [6:40 PM] 
We can have a counter and throw a warning.

[6:41] 
It won't break anything that is not already broken now, and it will probably fix any reasonable use case. (edited)

danielbachhuber [6:42 PM] 
I'm most concerned with restoring the pre-v1.2.0 behavior

schlessera [6:43 PM] 
That was basically "hope package combination and alphabetic file ordering do their jobs".

danielbachhuber [6:43 PM] 
yeah, not good enough :slightly_smiling_face:

schlessera [6:43 PM] 
I would guess that, with the above hook system, any existing packages will probably work without issues.

danielbachhuber [6:44 PM] 
if we don't fix it in v1.2.0, then we'll have to fix it with v1.2.1

schlessera [6:44 PM] 
This change is quickly implemented, I would suggest adding it into a branch and let the reporter of the above ticket run a test with their custom command.

danielbachhuber [6:45 PM] 
ok

[6:45] 
you should be able to test and reproduce, too

schlessera [6:45 PM] 
If it works as I think it does, it would keep all existing packages working, and be overall a reliable system to deal with ordering. (edited)

[6:46] 
> you should be able to test and reproduce, too
Yes, of course.

danielbachhuber [6:46 PM] 
ok

schlessera [6:47 PM] 
The basic algorithm is:
"if dependency is met -> add command; if dependency is not met -> hook up "add command" to "after-dependency-loaded"

danielbachhuber [6:48 PM] 
right. we just need the last step of "if dependency is never loaded, then register the command"

schlessera [6:48 PM] 
But that would just break...

danielbachhuber [6:48 PM] 
that's how the current behavior is though

[6:49] 
because the current system doesn't support declaring dependencies

[6:50] 
the current system is `WP_CLI::add_command( 'scaffold', 'Scaffold_Command' )` is run before packages are loaded, which means packages can call `WP_CLI::add_command( 'scaffold foo', 'Foo_Command' );` to register a `foo` subcommand in the `scaffold` command namespace

[6:50] 
if this works in v1.1.0, then it also needs to work (without changes) in v1.2.0 (edited)

schlessera [6:51 PM] 
Yes, and it would, no matter in what order both of these are loaded. (edited)

danielbachhuber [6:51 PM] 
right, but...

[6:51] 
the current system also supports calling `WP_CLI::add_command( 'foo bar', 'Bar_Command' );`, even if `foo` isn't ever registered.

[6:52] 
See `wp network meta` for an example of that

[6:52] 
so, v1.2.0 also needs to support registering a subcommand to a parent namespace, if the parent namespace never exists

[6:52] 
does this make sense?

schlessera [6:53 PM] 
Ah, I understand. Yes, makes sense.

[6:53] 
Do we have a hook that is triggered right after all commands have been loaded?

[6:53] 
`before_wp_load` ?

danielbachhuber [6:54 PM] 
`before_wp_load` probably isn't correct

[6:54] 
we can create a new hook for it

schlessera [6:54 PM] 
Maybe add a new one for this purpose then?

[6:54] 
Ok.

danielbachhuber [6:54 PM] 
yep

schlessera [6:54 PM] 
Any naming preference? `after_commands_loaded`?

danielbachhuber [6:54 PM] 
works for me

schlessera [6:55 PM] 
Do you want me to handle this, or will you do it?

danielbachhuber [6:55 PM] 
go for it :slightly_smiling_face:
```
